modules/3d/include/opencv2/3d/volume.hpp

```diff
@@ -5,124 +5,120 @@
-        pose(_pose),
-        raycastStepFactor(_raycastStepFactor)
-    { }
-
-    virtual ~Volume(){};
 
-    CV_WRAP
-    virtual void integrate(InputArray _depth, float depthFactor, const Matx44f& cameraPose,
-                           const Matx33f& intrinsics, const int frameId = 0)                   = 0;
-    virtual void integrate(InputArray _depth, InputArray _rgb, float depthFactor,
-                           const Matx44f& cameraPose, const Matx33f& intrinsics,
-                           const Matx33f& rgb_intrinsics, const int frameId = 0)               = 0;
-    CV_WRAP
-    virtual void raycast(const Matx44f& cameraPose, const Matx33f& intrinsics,
-                         const Size& frameSize, OutputArray points, OutputArray normals) const = 0;
-    virtual void raycast(const Matx44f& cameraPose, const Matx33f& intrinsics, const Size& frameSize,
-                         OutputArray points, OutputArray normals, OutputArray colors) const    = 0;
-    CV_WRAP
-    virtual void fetchNormals(InputArray points, OutputArray _normals) const                   = 0;
-    CV_WRAP
-    virtual void fetchPointsNormals(OutputArray points, OutputArray normals) const             = 0;
-    CV_WRAP
-    virtual void reset()                                                                       = 0;
-
-    // Works for hash-based volumes only, otherwise returns 1
-    virtual int getVisibleBlocks(int /*currFrameId*/, int /*frameThreshold*/) const { return 1; }
-    virtual size_t getTotalVolumeUnits() const { return 1; }
-
-   public:
-    const float voxelSize;
-    const float voxelSizeInv;
-    const Affine3f pose;
-    const float raycastStepFactor;
-};
-
-struct CV_EXPORTS_W VolumeParams
+class CV_EXPORTS_W Volume
 {
-    enum VolumeKind
-    {
-        TSDF        = 0,
-        HASHTSDF    = 1,
-        COLOREDTSDF = 2
-    };
-
-    /** @brief Kind of Volume
-        Values can be TSDF (single volume) or HASHTSDF (hashtable of volume units)
+public:
+    /** @brief Constructor of default volume - TSDF.
     */
-    CV_PROP_RW int kind;
-
-    /** @brief Resolution of voxel space
-        Number of voxels in each dimension.
-        Applicable only for TSDF Volume.
-        HashTSDF volume only supports equal resolution in all three dimensions
-    */
-    CV_PROP_RW int resolutionX;
-    CV_PROP_RW int resolutionY;
-    CV_PROP_RW int resolutionZ;
-
-    /** @brief Resolution of volumeUnit in voxel space
-        Number of voxels in each dimension for volumeUnit
-        Applicable only for hashTSDF.
+    Volume();
+    /** @brief Constructor of custom volume.
+    * @param vtype the volume type [TSDF, HashTSDF, ColorTSDF].
+    * @param settings the custom settings for volume.
     */
-    CV_PROP_RW int unitResolution = {0};
-
-    /** @brief Initial pose of the volume in meters, should be 4x4 float or double matrix */
-    CV_PROP_RW Mat pose;
+    Volume(VolumeType vtype, const VolumeSettings& settings);
+    ~Volume();
 
```

Comment:
I propose the following edits to integrate/raycast interface and docs:

1. Docs
1.1. No dots at the end of line
1.2. Now it's written that camera is in volume "environment", this is not true: the camera pos is in global coordinate system. This should be fixed.
1.3. Output points and normals are in camera's coordinate system. This also should be given in docs.
1.4. This is how I think this should be done:

for integrate:
```
    /** @brief Integrates the input data to the volume

    Camera intrinsics are taken from volume settings structure.

    * @param frame the object from which to take depth and image data
    * @param depth the depth image
    * @param color the color image (only for ColorTSDF)
    * @param pose the pose of camera in global coordinates
    */
```
for raycast:
```
    /** @brief Renders the volume contents into an image. The resulting points and normals are in camera's coordinate system.

    Rendered image size and camera intrinsics are taken from volume settings structure.

    * @param cameraPose the pose of camera in global coordinates
    * @param outFrame the object where to store rendered points and normals
    * @param points image to store rendered points
    * @param normals image to store rendered normals corresponding to points
    * @param colors image to store rendered colors corresponding to points (only for ColorTSDF).

    */
```

2. Interfaces:
```

    void raycast(InputArray cameraPose, OdometryFrame& outFrame) const;
    void raycast(InputArray cameraPose, OutputArray points, OutputArray normals) const;
    void raycast(InputArray cameraPose, OutputArray points, OutputArray normals, OutputArray colors) const;

```

Default argument `noArray()` should work fine with bindings: there are a lot of functions in OpenCV that work like that, see `Rodrigues()` for example.

Width and height should be taken from `getRaycastWidth/Height`, see the corresponding comment.